### PR TITLE
LPS-111216 Semver

### PR DESCRIPTION
**Pull request by adolfopa**
Hey @Preston-Crary 

This is a preliminary PR with some of the changes we talked about. All Depot tests pass locally (the CI will tell if this breaks anything).

I've taken a slightly different approach than the mapper we talked about initially. I abandoned the mapper idea because I wanted this extension point to support all existing cases; this sometimes requires more than simple role membership checking. For example, in `isGroupAdminImpl` we have the following:

```
else if (group.isLayoutPrototype()) {
	if (LayoutPrototypePermissionUtil.contains(
			this, group.getClassPK(), ActionKeys.UPDATE)) {

		return true;
	}

	return false;
}
```

With the approach in this PR -- having 4 membership predicates instead of a mapper -- we can cover all current usages. So, in theory, we could move all that layout/organization/etc. specific logic into membership customizers.

The only case that the mapper covers which is missing in the PR is the key used when caching the result. But the thing is that the key doesn't really correspond with a real assignment anyway (e.g. a SITE_ADMIN that doesn't have the content reviewer role will have its result cached under the reviewer role key anyway); in fact it could be anything, as long as it is used consistently. So, given that this is an implementation detail -- how AdvPermissionChecker does the caching -- I think we can leave it as it. Alternatively, we could use a different key that does not refer to sites. E.g. (I've simply replaced the references to the RoleConstants.SITE_CONTENT_REVIEWER)
```
Boolean value = PermissionCacheUtil.getUserPrimaryKeyRole(
	getUserId(), groupId, RoleCacheConstants.GROUP_CONTENT_REVIEWER); 

try {
	if (value == null) {
		value = isContentReviewerImpl(groupId);

		PermissionCacheUtil.putUserPrimaryKeyRole(
			getUserId(), groupId, RoleCacheConstants.GROUP_CONTENT_REVIEWER,
			value);
	}
}
catch (Exception exception) {
	PermissionCacheUtil.removeUserPrimaryKeyRole(
		getUserId(), groupId, RoleCacheConstants.GROUP_CONTENT_REVIEWER);

	throw exception;
}
```

Permission checking is still handled by the wrapper. Should we add a similar extension point for that? I think it is better to keep role checking and permission checking separate; these are different things, and not everybody will need to override/extend both at the same time. So if we want to move that logic out of the wrapper, I believe it would be better to add an additional extension point.

Finally, I'm not happy with the naming, but I couldn't come up with anything better. Feel free to change anything you want.

Let me know what do you think about this.

Thanks!

### CODE REVIEW

**Comment by adolfopa** [comment]
ci:test:sf

**Comment by adolfopa** [comment]
ci:test:relevant

File: modules/apps/depot/depot-service/src/main/java/com/liferay/depot/internal/security/permission/DepotGroupMembershipCustomizer.java [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+
+			return true;
+		}
+
+		return false;
+	}
+
+	@Override
+	public Boolean isGroupContentReviewer(User user, Group group) {
+		return null;
```
Comment (adolfopa) [inline modules/apps/depot/depot-service/src/main/java/com/liferay/depot/internal/security/permission/DepotGroupMembershipCustomizer.java]: I'm returning null on purpose; the default logic is enough.

File: modules/apps/depot/depot-service/src/main/java/com/liferay/depot/internal/security/permission/DepotGroupMembershipCustomizer.java [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+	property = "group.class.name=com.liferay.depot.model.DepotEntry",
+	service = GroupMembershipCustomizer.class
+)
+public class DepotGroupMembershipCustomizer
+	implements GroupMembershipCustomizer {
+
+	@Override
+	public Boolean isGroupAdmin(User user, Group group) throws PortalException {
+		if (group.getType() != GroupConstants.TYPE_DEPOT) {
+			return false;
```
Comment (adolfopa) [inline modules/apps/depot/depot-service/src/main/java/com/liferay/depot/internal/security/permission/DepotGroupMembershipCustomizer.java]: This should be null, not false (sorry!)

Comment (Preston-Crary) [reply modules/apps/depot/depot-service/src/main/java/com/liferay/depot/internal/security/permission/DepotGroupMembershipCustomizer.java]: Haha, yeah I figured that.  I'll update it.

**Comment by Preston-Crary** [comment]
Please see https://github.com/csierra/liferay-portal/pull/663

**Pull request closed without merging**